backend/app/usecase/url/urlupdater.go

```diff
@@ -0,0 +1,99 @@
+package url
+
+import (
+	"github.com/short-d/app/fw/timer"
+	"github.com/short-d/short/backend/app/entity"
+	"github.com/short-d/short/backend/app/usecase/keygen"
+	"github.com/short-d/short/backend/app/usecase/repository"
+	"github.com/short-d/short/backend/app/usecase/risk"
+	"github.com/short-d/short/backend/app/usecase/validator"
+)
+
+var _ Updater = (*UpdaterPersist)(nil)
+
+// Updater represents a short link attribute updater.
+type Updater interface {
+	UpdateURL(oldAlias string, update entity.URL, user entity.User) (entity.URL, error)
+}
+
+// UpdaterPersist represents a short link updater that persists the given changes
+type UpdaterPersist struct {
+	urlRepo             repository.URL
+	userURLRelationRepo repository.UserURLRelation
+	keyGen              keygen.KeyGenerator
+	longLinkValidator   validator.LongLink
+	aliasValidator      validator.CustomAlias
+	timer               timer.Timer
+	riskDetector        risk.Detector
+}
+
+// UpdateURL persists mutations for a given short link in the repository.
+func (u UpdaterPersist) UpdateURL(
+	oldAlias string,
+	update entity.URL,
+	user entity.User,
+) (entity.URL, error) {
+	url, err := u.urlRepo.GetByAlias(oldAlias)
+
+	if err != nil {
+		return entity.URL{}, err
+	}
+
+	url = u.updateAlias(url, update)
+	url = u.updateLongLink(url, update)
+
+	if u.riskDetector.IsURLMalicious(url.OriginalURL) {
+		return entity.URL{}, ErrMaliciousLongLink(url.OriginalURL)
+	}
+
+	if !u.aliasValidator.IsValid(&url.Alias) {
+		return entity.URL{}, ErrInvalidCustomAlias(url.Alias)
+	}
+
+	if !u.longLinkValidator.IsValid(&url.OriginalURL) {
+		return entity.URL{}, ErrInvalidLongLink(url.OriginalURL)
+	}
```

Comment:
Before persisting the updates yes, but not before merging the changes since we cannot predict if the user will only update the short link and give a blank long link. That was my though process when working through the solution.

Reply:
@oatovar Makes sense.

Reply:
@oatovar Can we move validation in front of risk detection? Risk detection is much more expensive because it invokes network calls.